arpsync/watcher.go

```diff
@@ -0,0 +1,190 @@
+	containersMap := make(map[string]*metadata.Container)
+
+	for index, aContainer := range containers {
+		if !(aContainer.PrimaryIp != "" && aContainer.NetworkUUID == network.UUID) {
+			continue
+		}
+		containersMap[aContainer.PrimaryIp] = &containers[index]
+	}
+
+	return containersMap, nil
+}
+
+func (atw *ARPTableWatcher) syncLoop() {
+
+	logrus.Infof("arpsync: starting monitoring every %v seconds", atw.syncInterval)
+	for {
+		time.Sleep(atw.syncInterval)
+		logrus.Debugf("arpsync: time to sync ARP table")
+		err := atw.doSync()
+		if err != nil {
+			logrus.Errorf("arpsync: while syncing, got error: %v", err)
+		}
+	}
+}
+
+func (atw *ARPTableWatcher) doSync() error {
+	logrus.Debugf("arpsync: checking the ARP table")
+	networks, err := atw.mc.GetNetworks()
+	if err != nil {
+		logrus.Errorf("arpsync: error fetching networks from metadata")
+		return err
+	}
+
+	host, err := atw.mc.GetSelfHost()
+	if err != nil {
+		logrus.Errorf("arpsync: error fetching self host from metadata")
+		return err
+	}
+
+	services, err := atw.mc.GetServices()
+	if err != nil {
+		logrus.Errorf("arpsync: error fetching services from metadata")
+		return err
+	}
+
+	var networkDriverMacAddress string
+	localNetworks := map[string]bool{}
+	for _, service := range services {
+		// Trick to select the primary service of the network plugin
+		// stack
+		// TODO: Need to check if it's needed for Calico?
+		if !(service.Kind == "networkDriverService" &&
+			service.Name == service.PrimaryServiceName) {
+			continue
+		}
+
+		logrus.Debugf("arpsync: service: %#v", service)
+		for _, aContainer := range service.Containers {
+			if aContainer.HostUUID == host.UUID {
+				networkDriverMacAddress = aContainer.PrimaryMacAddress
+				localNetworks[aContainer.NetworkUUID] = true
+			}
+		}
+	}
+	if len(localNetworks) == 0 {
+		return fmt.Errorf("couldn't find any local networks")
+	}
+	logrus.Debugf("arpsync: localNetworks: %v", localNetworks)
+	logrus.Debugf("arpsync: networkDriverMacAddress=%v", networkDriverMacAddress)
+
+	var localNetwork metadata.Network
+	for _, aNetwork := range networks {
+		if _, ok := localNetworks[aNetwork.UUID]; ok {
+			localNetwork = aNetwork
+			break
+		}
+	}
+	logrus.Debugf("arpsync: localNetwork: %+v", localNetwork)
+
+	// Read the ARP table
+	entries, err := netlink.NeighList(0, netlink.FAMILY_V4)
+	if err != nil {
+		logrus.Errorf("arpsync: error fetching entries from ARP table")
+		return err
+	}
+	logrus.Debugf("arpsync: entries=%+v", entries)
+
+	containers, err := atw.mc.GetContainers()
+	if err != nil {
+		logrus.Errorf("arpsync: error fetching containers from metadata")
+		return err
+	}
+	containersMap, err := buildContainersMap(containers, localNetwork)
+	//logrus.Debugf("arpsync: containersMap: %v", containersMap)
+
+	for _, aEntry := range entries {
+		//logrus.Debugf("arpsync: aEntry: %+v", aEntry)
+		if container, found := containersMap[aEntry.IP.String()]; found {
+			if container.HostUUID == host.UUID {
```

Comment:
it appears this if/else is just copy and paste but the desired mac address is different.  Can you simplify the code.